simvue/eco/emissions_monitor.py

```diff
@@ -0,0 +1,325 @@
+        self._logger = _logger
+        self._client: APIClient = APIClient(
+            co2_api_token=self.co2_signal_api_token, timeout=10
+        )
+        self._processes: dict[str, ProcessData] = {}
+
+    @pydantic.validate_call(config={"arbitrary_types_allowed": True})
+    def attach_process(
+        self, process: psutil.Process | None = None, label: str | None = None
+    ) -> str:
+        """
+        Attach a process to the CO2 Monitor.
+
+        Parameters
+        ----------
+        process : psutil.Process | None
+            The process to monitor, if None measures the current running process. Default is None.
+        label : str | None
+            The label to assign to the process. Default is process_<pid>.
+
+        Returns
+        -------
+        int
+            The PID of the process.
+        """
+        if process is None:
+            process = psutil.Process(pid=os.getpid())
+
+        self._logger.info(f"📎 Attaching process with PID {process.pid}")
+
+        label = label or f"process_{process.pid}"
+        self._processes[label] = ProcessData(process=process)
+
+        return label
+
+    def estimate_co2_emissions(self) -> None:
+        """Estimate the CO2 emissions"""
+        self._logger.info("📐 Measuring CPU usage and power.")
+
+        if self._local_data is None:
+            raise RuntimeError("Expected local data to be initialised.")
+
+        if not self._data_file_path:
+            raise RuntimeError("Expected local data file to be defined.")
+
+        if (
+            not self.co2_intensity
+            and not self._local_data.setdefault(self._client.country_code, {})
+            or self.outdated
+        ):
+            self._logger.info("🌍 CO2 emission outdated, calling API.")
+            _data: CO2SignalResponse = self._client.get()
+            self._local_data[self._client.country_code] = _data.model_dump(mode="json")
+            self._local_data["last_updated"] = self.now()
+
+            with self._data_file_path.open("w") as out_f:
+                json.dump(self._local_data, out_f, indent=2)
+
+        if self.co2_intensity:
+            _current_co2_intensity = self.co2_intensity
+            _co2_units = "kgCO2/kWh"
+        else:
+            self._current_co2_data = CO2SignalResponse(
+                **self._local_data[self._client.country_code]
+            )
+            _current_co2_intensity = self._current_co2_data.data.carbon_intensity
+            _co2_units = self._current_co2_data.carbon_intensity_units
+
+        for label, process in self._processes.items():
+            process.cpu_percentage = process.process.cpu_percent(
+                interval=self.cpu_interval
+            )
+            _previous_energy: float = process.total_energy
+            process.power_usage = min(
+                self.cpu_idle_power,
+                (process.cpu_percentage / 100.0) * self.thermal_design_power_per_core,

```

Comment:
Will be helpful to document this. cpu_percent is the percentage of CPU time the process used during the interval between the last call and current call. So 20% implies that the power usage is (20/100)*85W = 17W

Reply:
I have now removed the "idle power" because we do not care about this for relative comparison and does not represent resource usage for the simulation itself.

Reply:
Added in: https://github.com/simvue-io/docs/pull/61